cirq/google/op_deserializer_test.py

```diff
@@ -362,3 +329,121 @@ def test_token_with_references():
+    op1.qubits.add().id = '1_1'
+
+    return v2.program_pb2.Circuit(
+        scheduling_strategy=v2.program_pb2.Circuit.MOMENT_BY_MOMENT,
+        moments=[
+            v2.program_pb2.Moment(
+                operations=[op1],
+            ),
+        ],
+    )
+
+
+def default_circuit():
+    return cirq.FrozenCircuit(
+        cirq.X(cirq.GridQubit(1, 1)) ** sympy.Symbol('k'),
+        cirq.measure(cirq.GridQubit(1, 1), key='m'),
+    )
+
+
+def test_circuit_op_from_proto_errors():
+    deserializer = cg.CircuitOpDeserializer()
+    serialized = v2.program_pb2.CircuitOperation(circuit_constant_index=0)
+
+    constants = [v2.program_pb2.Constant(circuit_value=default_circuit_proto())]
+    raw_constants = [default_circuit()]
+
+    with pytest.raises(ValueError, match='CircuitOp deserialization requires a constants list'):
+        deserializer.from_proto(serialized)
+
+    with pytest.raises(ValueError, match='CircuitOp deserialization requires a constants list'):
+        deserializer.from_proto(serialized, constants=constants)
+
+    with pytest.raises(ValueError, match='CircuitOp deserialization requires a constants list'):
+        deserializer.from_proto(serialized, raw_constants=raw_constants)
+
+    bad_raw_constants = [2]
+    with pytest.raises(ValueError, match='Constant at index 0 was expected to be a circuit'):
+        deserializer.from_proto(serialized, constants=constants, raw_constants=bad_raw_constants)
+
+
+def test_circuit_op_arg_key_errors():
+    deserializer = cg.CircuitOpDeserializer()
+    arg_map = v2.program_pb2.ArgMapping()
+    p1 = arg_map.entries.add()
+    p1.key.arg_value.float_value = 1.0
+    p1.value.arg_value.float_value = 2.0
+
+    serialized = v2.program_pb2.CircuitOperation(circuit_constant_index=0, arg_map=arg_map)
+
+    constants = [v2.program_pb2.Constant(circuit_value=default_circuit_proto())]
+    raw_constants = [default_circuit()]
+
+    with pytest.raises(ValueError, match='Invalid key parameter type'):
+        deserializer.from_proto(serialized, constants=constants, raw_constants=raw_constants)
+
+
+def test_circuit_op_arg_val_errors():
+    deserializer = cg.CircuitOpDeserializer()
+    arg_map = v2.program_pb2.ArgMapping()
+    p1 = arg_map.entries.add()
+    p1.key.arg_value.string_value = 'k'
+    p1.value.arg_value.bool_values.values.extend([True, False])
+
+    serialized = v2.program_pb2.CircuitOperation(circuit_constant_index=0, arg_map=arg_map)
+
+    constants = [v2.program_pb2.Constant(circuit_value=default_circuit_proto())]
+    raw_constants = [default_circuit()]
+
+    with pytest.raises(ValueError, match='Invalid value parameter type'):
+        deserializer.from_proto(serialized, constants=constants, raw_constants=raw_constants)
+
+
+def test_circuit_op_from_proto():
+    deserializer = cg.CircuitOpDeserializer()
+
+    repetition_spec = v2.program_pb2.RepetitionSpecification()
+    repetition_spec.repetition_count = 1
```

Comment:
Should we have a test with multiple repetitions and/or repetition ids?

Reply:
Done. No deserializer update was needed (unlike the serializer).